tool/src/dart/mod.rs

```diff
@@ -248,34 +247,121 @@ impl<'a, 'cx> TyGenContext<'a, 'cx> {
                     annotation,
                     ffi_cast_type_name,
                     dart_type_name,
-                    get_expression,
-                    set_expressions,
+                    c_to_dart,
+                    dart_to_c,
+                    ty: &field.ty,
                 }
             })
             .collect::<Vec<_>>();
 
-        let methods = ty
+        let mut methods = ty
             .methods
             .iter()
             .flat_map(|method| self.gen_method_info(id, method, type_name))
             .collect::<Vec<_>>();
 
+        // Non-out structs needs to be constructible in Dart
+        let default_constructor = if !is_out {
+            if let Some(constructor) = methods
+                .iter_mut()
+                .find(|m| m.declaration.contains(&format!("{type_name}()")))
+            {
+                // If there's an existing zero-arg constructor, we repurpose it with optional arguments for all fields
+                let mut is_first = true;
+                let args = fields.iter().fold(String::new(), |acc, field| {
+                    format!(
+                        "{acc}{comma}{typ}? {name}",
+                        comma = if is_first {
+                            is_first = false;
+                            ""
+                        } else {
+                            ", "
+                        },
+                        typ = field.dart_type_name,
+                        name = field.name
+                    )
+                });
+                constructor.declaration = format!("factory {type_name}({{{args}}})");
+
+                let mut r = String::new();
+                writeln!(&mut r, "final dart = {type_name}._(result);").unwrap();
+                for field in &fields {
+                    let name = &field.name;
+                    writeln!(&mut r, "if ({name} != null) {{").unwrap();
+                    writeln!(&mut r, "  dart.{name} = {name};").unwrap();
+                    writeln!(&mut r, "}}").unwrap();
+                }
+                write!(&mut r, "return dart;").unwrap();
+                constructor.return_expression = Some(r.into());
+
+                None
+            } else {
+                // Otherwise we create a constructor with default values for all fields.
+                let mut is_first = true;
+                let args = fields.iter().fold(String::new(), |acc, field| {
+                    let defaultt: Cow<'static, str> = match &field.ty {
```

Comment:
I'm torn between using default values and requiring all arguments. It's more ergonomic to have default values, and it should be fine to use the values that produce the all-zero byte pattern, because if we calloced the struct we'd get the same.

That said, some of these might be weird defaults.

Reply:
Imo we should require all arguments.

A thing I would like to support in the long run is *enum* defaults so that you can e.g. pass a partial dict in JS or Dart and get stuff converted.

Basically the default behavior can be opt in and will only work for default able fields.

Reply:
I'm already allowing defaults for all fields, like here: https://github.com/rust-diplomat/diplomat/pull/408/files#diff-0ce623e647a74cad01dbf61059bd0c817bafc73cd6209385e08239123496a4cbR32

Reply:
I mean defaults that can be obtained without calling FFI. That would avoid needing to jump the boundary twice